src/jose.c

```diff
@@ -922,21 +923,51 @@ static apr_byte_t oidc_jose_zlib_uncompress(apr_pool_t *pool, const char *input,
 	zlib.opaque = Z_NULL;
 	zlib.avail_in = (uInt) input_len;
 	zlib.next_in = (Bytef*) input;
-	*output = apr_pcalloc(pool, input_len * 4);
-	zlib.avail_out = (uInt) (input_len * 4);
-	zlib.next_out = (Bytef*) (*output);
+	zlib.total_out = 0;
+	size_t uncompLength = OIDC_CJOSE_UNCOMPRESS_CHUNK;
+	char *uncomp = (char *) calloc(sizeof (char), uncompLength + 1);
+	if (uncomp == NULL) {
+		oidc_jose_error(err, "memory allocation error");
+		return FALSE;
+	}
 
 	inflateInit(&zlib);
-	if (inflate(&zlib, Z_FINISH) != Z_STREAM_END) {
-		oidc_jose_error(err, "inflate failed");
-		return FALSE;
+	int done = 1;
+	while (done) {
+		int inflate_status;
+		if (zlib.total_out >= OIDC_CJOSE_UNCOMPRESS_CHUNK) {
+			char *uncomp2 = (char *) calloc(sizeof (char), uncompLength + OIDC_CJOSE_UNCOMPRESS_CHUNK);
+			if (uncomp2 == NULL) {
+				free(uncomp);
+				oidc_jose_error(err, "memory allocation error");
+				return FALSE;
+			}
+			_oidc_memcpy(uncomp2, uncomp, uncompLength);
+			uncompLength += OIDC_CJOSE_UNCOMPRESS_CHUNK;
+			free(uncomp);
+			uncomp = uncomp2;
+		}
+		zlib.next_out = (Bytef *) (uncomp + zlib.total_out);
+		zlib.avail_out = (uInt) uncompLength - zlib.total_out;
+		inflate_status = inflate(&zlib, Z_SYNC_FLUSH);
+		if (inflate_status == Z_STREAM_END) {
+			done = 0;
+		} else if (inflate_status != Z_OK) {
+			oidc_jose_error(err, "inflate failed. ret:[%d]", inflate_status);
+			free(uncomp);
+			return FALSE;
+		}
 	}
+
 	if (inflateEnd(&zlib) != Z_OK) {
 		oidc_jose_error(err, "inflateEnd failed");
+		free(uncomp);
 		return FALSE;
 	}
 
 	*output_len = (int) zlib.total_out;
+	*output = (char *)apr_pstrdup(pool, uncomp);
```

Comment:
though we know that compression right now is only used with string payloads, it would be more appropriate to use _oidc_memcpy here